src/inner.rs

```diff
@@ -0,0 +1,224 @@
+//! The inner mechanism powering the `Event` type.
+
+use crate::list::{Entry, List};
+use crate::node::Node;
+use crate::queue::Queue;
+use crate::sync::atomic::{AtomicBool, AtomicUsize, Ordering};
+use crate::sync::cell::UnsafeCell;
+use crate::Task;
+
+use alloc::vec;
+use alloc::vec::Vec;
+
+use core::ops;
+use core::ptr::NonNull;
+
+/// Inner state of [`Event`].
+pub(crate) struct Inner {
+    /// The number of notified entries, or `usize::MAX` if all of them have been notified.
+    ///
+    /// If there are no entries, this value is set to `usize::MAX`.
+    pub(crate) notified: AtomicUsize,
+
+    /// A linked list holding registered listeners.
+    list: Mutex<List>,
+
+    /// Queue of nodes waiting to be processed.
+    queue: Queue,
+
+    /// A single cached list entry to avoid allocations on the fast path of the insertion.
+    cache: UnsafeCell<Entry>,
```

Comment:
document how access to this field is locked when necessary. (that is, what is the invariant under which that field can be modified and/or read)